### ISSUE TITLE
TWS version

### ISSUE THREAD

**Issue by heshriti** [IbcAlpha/IBC#126]
Hi,

I have an older version of gateway installed, and have not used it in a while.
When I tried logging, it errored out saying the 963 version is not supported.

Is it possible to get it working without an upgrade? 
Or moving to the latest version is the only option?

thanks.

**Comment by rlktradewright** [comment]
No it's not possible, just upgrade.

It will only take you a few minutes to download and install the current Gateway.

If you're still using an old version of IBC, you'll need to upgrade that as well because there were some breaking changes in the login dialog in a recent Gateway version (980).

**Comment by heshriti** [comment]
Thanks, Richard. I will do that.
I was using IBcontroller before. 

Per  your docs, 972+ versions need to be restarted daily.
Do we have control on when IB is shut  down daily? It seems the US users will  be shut off around midnight,  effectively shutting them off from many markets.

**Comment by rlktradewright** [comment]
You have complete control over when TWS/Gateway do their daily shut down (this has been the case forever). Look in the 'Lock and Exit' section of the global configuration dialog.

I shut mine down around 05:40 UK time, and restart them a couple of minutes later. At this time of day there doesn't seem to be much happening anywhere that is of interest to me, but of course others have different trading interests and will need to determine what's best for them.

If someone really can't accept even a couple of minutes downtime, then the only solution is to not use IBC and instead use the auto-relogin facilities included in TWS/Gateway 974 and later.

**Comment by heshriti** [comment]
That's interesting. 
The last I used IB- it  used to shut down for 45 mins to 1.5 hours or so around midnight-- and I tried various things like change the time zone in the files. Didnt work.

**Comment by heshriti** [comment]
After considerable struggle, I was able to get it working. 
One nuisance is every time I log in gate way displays that "US bid ask is now quoted in shares". I have accepted that check box and asked not to repeat. But it repeats itself again when I tried to log in?
ANy way I can switch off this dialog box at startup?

**Comment by rlktradewright** [comment]
1. The TWS autologoff is not the same thing as the IB nightly server reset, which is what I presume you're referring to when you say 'it used to shut down for 45 mins to 1.5 hours or so around midnight'. There is nothing you can do to prevent or change the time of the nightly server reset (apart from having your IB account hosted in a different region, since different regions have different reset periods.) And I've never experienced the nightly reset being anything like as long as 1.5 hours: these days it's just a few minutes.

   The auto-logoff mechanism has been there in TWS since way before I started using it in 2003, and indeed is part of the reason that IBController was originally developed: it kept TWS running by changing the auto-logoff time about 5 minutes before the currently defined time was reached, so that it never actually was reached. But that workaround no longer works. You can set the autologoff time to any time you like. The autologoff mechanism was added to the Gateway in version 974: before that the Gateway used to just keep running until you shut it down: but those days are gone...

2. You say: 'After considerable struggle, I was able to get it working'. Are you talking about IBC or TWS 963? If you mean IBC, I'd be interested to know what you struggled with, because I've made huge efforts over the years to make IBC as easy to install as possible.

3. I'm not familiar with the message that says "US bid ask is now quoted in shares", probably because I saw it once, asked it not to show again, and have never seen it since. I've no idea why it's still showing it to you, unless you're wiping out the Gateway settings file after you shut it down.

4. You can stop IBC tidily by sending a STOP command to the IBC command server using telnet. There are settings related to the command server in config.ini, so make sure you set those as appropriate (by default the command server is not configured and does not start). So connect to the command server port (using telnet or any other means of establishing a socket connection) and then just send the text 'STOP' and IBC will immediately shut down tidily.

**Comment by heshriti** [comment]
Thanks Richard. I appreciate  your detailed clarifications and responses.

On 2)  I upgraded IB to the current official version. There were many issues with my headless setup- IBC and not IBC related. As for the IBC installation,  I originally downloaded the git repo on my local dir-path, and setup the appropriate paths/variables/environment and kept trying to run the (tws)gatewaystart script. It just returned nothing. All it would do is return nothing - sort of like echo "" behavior. I was wondering if this was due to my headless setup . 
So I manually went through the scripts folder, and tried breakpoints and print statements within displaybanner and ibcstart scripts. That didnt go very far. I then re-read the official documentation- and found that the ibc*.zip needs to be unzipped in /opt/ibc directory.  

So I did that- and then everything started to work. :)  

I guess something was missing in the enviroment variables that if you just download/extract in your local folder, the install fails.

**Comment by rlktradewright** [comment]
IBC doesn't _have_ to be installed in /opt/ibc. You can install it anywhere you like.

I just put it on my Ubuntu 20.04 VM: I extracted the zip to a new empty folder ~/Qaz, set the execute permissions on the scripts, and it ran straight away using TWS 983. I didn't have to do anything else (except set the username and password in config.ini). It really couldn't have been any easier!

If I had a different TWS installed, all I would have had to do was set the correct version number in twsstart.sh.

So no, there is nothing missing in the environment variables, and if you download and extract in your local folder it works just fine.

**Comment by heshriti** [comment]
That's what I thought and have always assumed.  somehow it didnt work for me. 

Also,  is it possible to encrypt pasword in the file. In IBController-- I believe something like this:
java -cp IBController.jar ibcontroller.IBController encrypt pwd

Does IBC have this?

**Comment by rlktradewright** [comment]
No, I removed the password 'encryption' mechanism because it was practically useless and also failed to work if the password contained certain characters.

**Comment by rlktradewright** [comment]
By the way, there was a discussion about the password 'encryption' on the IBC User Group last year, which explains my reasoning in more depth. See https://groups.io/g/ibcalpha/topic/75516160#1161
